### DISCUSSION TITLE
CommonJS issue with Styled Components on Build

### DISCUSSION THREAD

**Discussion by anujshah108** [vitejs/vite-plugin-react#383] [Q&A]
Hi all -

I am getting an issue when building my app that I recently converted from CRA to VITE. 

`error during build:
node_modules/styled-components/dist/styled-components.browser.esm.js (1:9): "typeOf" is not exported by "node_modules/react-is/index.js", imported by "node_modules/styled-components/dist/styled-components.browser.esm.js".`

This only happens on build and I have tried all the commonJS options and plugins and nothing seems to fix it.

Please let me know!

**Comment by anujshah108** [comment]
This is my vite.config.ts:

````
export default defineConfig({
  define: {
    "process.env": {},
    globalThis: "window",
    global: "globalThis",
  },
  optimizeDeps: {
    include: ["linked-dep"],
  },
  resolve: {
    alias: {
      // This Rollup aliases are extracted from @esbuild-plugins/node-modules-polyfill,
      // see https://github.com/remorses/esbuild-plugins/blob/master/node-modules-polyfill/src/polyfills.ts
      // process and buffer are excluded because already managed
      // by node-globals-polyfill
      util: "rollup-plugin-node-polyfills/polyfills/util",
      sys: "util",
      events: "rollup-plugin-node-polyfills/polyfills/events",
      stream: "rollup-plugin-node-polyfills/polyfills/stream",
      path: "rollup-plugin-node-polyfills/polyfills/path",
      querystring: "rollup-plugin-node-polyfills/polyfills/qs",
      punycode: "rollup-plugin-node-polyfills/polyfills/punycode",
      url: "rollup-plugin-node-polyfills/polyfills/url",
      string_decoder: "rollup-plugin-node-polyfills/polyfills/string-decoder",
      http: "rollup-plugin-node-polyfills/polyfills/http",
      https: "rollup-plugin-node-polyfills/polyfills/http",
      dwt: "rollup-plugin-node-polyfills/polyfills/os",
      assert: "rollup-plugin-node-polyfills/polyfills/assert",
      constants: "rollup-plugin-node-polyfills/polyfills/constants",
      _stream_duplex:
        "rollup-plugin-node-polyfills/polyfills/readable-stream/duplex",
      _stream_passthrough:
        "rollup-plugin-node-polyfills/polyfills/readable-stream/passthrough",
      _stream_readable:
        "rollup-plugin-node-polyfills/polyfills/readable-stream/readable",
      _stream_writable:
        "rollup-plugin-node-polyfills/polyfills/readable-stream/writable",
      _stream_transform:
        "rollup-plugin-node-polyfills/polyfills/readable-stream/transform",
      timers: "rollup-plugin-node-polyfills/polyfills/timers",
      console: "rollup-plugin-node-polyfills/polyfills/console",
      vm: "rollup-plugin-node-polyfills/polyfills/vm",
      zlib: "rollup-plugin-node-polyfills/polyfills/zlib",
      tty: "rollup-plugin-node-polyfills/polyfills/tty",
      domain: "rollup-plugin-node-polyfills/polyfills/domain",
    },
  },

  // This changes the out put dir from dist to build
  // comment this out if that isn't relevant for your project
  build: {
    outDir: "build",
    sourcemap: true,
    commonjsOptions: {
      include: [/linked-dep/, /node_modules/],
    },
    rollupOptions: {
      // make sure to externalize deps that shouldn't be bundled
      // into your library
      external: ["react"],
      output: {
        // Provide global variables to use in the UMD build
        // for externalized deps
        globals: {
          react: "React",
        },
      },
    },
  },
  test: {
    globals: true,
    environment: "jsdom",
    setupFiles: "./setupTest.ts",
    // you might want to disable it, if you don't have tests that rely on CSS
    // since parsing CSS is slow
    css: true,
  },

  plugins: [
    legacy({
      targets: [
        "chrome >= 64",
        "edge >= 79",
        "safari >= 11.1",
        "firefox >= 67",
      ],
      renderLegacyChunks: false,
      modernPolyfills: ["es/global-this"],
    }),
    react({
      babel: {
        plugins: ["styled-components"],
        babelrc: false,
        configFile: false,
      },
    }),
    checker({
      typescript: true,
      eslint: {
        lintCommand: 'eslint "./src/**/*.{ts,tsx}"',
      },
    }),
    macrosPlugin(),
    // eslint(),
    nodePolyfills(),
    NodeGlobalsPolyfillPlugin(),
    NodeModulesPolyfillPlugin(),
    babel({
      babelConfig: {
        babelrc: false,
        compact: false,
        configFile: false,
      },
    }),
    sentryVitePlugin({
      org: "billion-to-one",
      project: "patient-portal-frontend",
      authToken: process.env.VITE_SENTRY_AUTH_TOKEN,
    }),
  ],
});`
```

**Reply by anujshah108** [reply]
solved by adding to react-is to external roll-up options

**Reply by ArnaudBarre** [reply]
Nice you manage to solve this. And sorry I didn't answer sooner.
Are you building an app or a lib?
If it's an app, react should not be externalized
If it's a lib, you may want to switch to another tool for [advance usecases](https://vite.dev/guide/build#advanced-base-options:~:text=bundlers%20and%20runtimes.-,advanced%20usage,-Library%20mode%20includes)

For information, Vite philosophy is to use modern dependency that have browser bundles and don't expect node globals to be present. While using node polyfills works for now, it's highly recommended to move out from that, it will make your configuration much more simple.

**Reply by anujshah108** [reply]
Thank you!

I will remove the external of react -

I am now getting this error unfortunetly:

`"default" is not exported by "node_modules/prop-types/index.js"`

I have the prop-types types library installed - so not sure how to approach this.

Please let me know if you know anything offhand

**Reply by ArnaudBarre** [reply]
It might be an issue in Vite for esm/cjs compatibility. Prop types have been deprecated for a long time. All the react libs I used have now published versions without a prop types dependence.

You can probably find people that found workaround like a patch to fix support for prop types otherwise

**Reply by anujshah108** [reply]
Thank you! I upgraded material and that worked! Last question :) (promise) ! I am getting this now randomly:

`node_modules/styled-components/dist/styled-components.browser.esm.js (2:7): "default" is not exported by "node_modules/react/index.js", imported by "node_modules/styled-components/dist/styled-components.browser.esm.js".`

**Reply by ArnaudBarre** [reply]
This is yet again an cjs/esm interop issue. Given that `styled-components` is very popular you're probably not the first one to run into it. Are you using the latest version? If yes I can look into it, this might be an issue on how they bundle their lib

**Reply by anujshah108** [reply]
I am using the latest yes and just added this to the vite.config:

```
 react({
      babel: {
        plugins: [
          [
            "babel-plugin-styled-components",
            {
              ssr: false,
              pure: true,
              displayName: true,
              fileName: false,
            },
          ],
        ],
      },
    }),

```
 
The cjs error went away. I am already using the macros plugin but now I get this error now but ONLY on build - running the app works fine:


`Cannot find module 'styled-components/macro' from`

**Reply by ArnaudBarre** [reply]
Yeah this one of the biggest issue in Vite that Rolldown will solve. Dependency are bundle with esbuild in dev and Rollup for builds, and they don't handle CJS to ESM exactly the same way.

That's strange that the import is not resolved, but probably you don't want to use that, because the macro seems to be mainly to workaround the fact that in CRA you can't add babel plugins. Using the babel plugin as you did should remove the need for the macro.

**Reply by anujshah108** [reply]
Thanks yeah I went for that solution - that seemed to work and the styles are still valid.

I am getting this error now (again only on Build) :(. Hopefully this is the last thing. Seems to be a CJS issue again. I see online people used namedExports with commonjs to solve it before but that doesn't seem to be an option anymore?

`"useRef" is not exported by "node_modules/react/index.js"`

**Reply by ArnaudBarre** [reply]
The default template uses named export and this is not an issue.
Maybe this is because of the legacy plugin? or another plugin in your config. Try either removing them or starting from a fresh install, copy your src code in it and then try to add only the necessary parts to make it works, it can help you avoid bringing legacy from CRA

**Reply by anujshah108** [reply]
Thanks - I tried all the above including just porting over to a fresh install. Same error. Let me know if you have any other tips. Appreciate all the help nonetheless!

**Reply by ArnaudBarre** [reply]
Can you share start from https://vite.new/react-ts and the shortest possible number of line to repro the issue? I'm really surprised that you get issue for React itself

**Reply by anujshah108** [reply]
I was able to get this working by deleting all node_modules and then re-add each package one by one and then also simplify the vite config. 

I have no idea what worked tbh but its working for now.

Thanks for all the help!

**Reply by ArnaudBarre** [reply]
Happy to see you've solved it and succeeded at reducing tech debt!
